src/wh_server_keystore.c

```diff
@@ -0,0 +1,296 @@
+#include "wolfssl/wolfcrypt/error-crypt.h"
+
+#include "wolfhsm/wh_server.h"
+#include "wolfhsm/wh_server_keystore.h"
+#include "wolfhsm/wh_message.h"
+#include "wolfhsm/wh_packet.h"
+#include "wolfhsm/wh_error.h"
+
+int hsmGetUniqueId(whServerContext* server)
+{
+    int i;
+    int ret = 0;
+    whNvmId id;
+    whNvmId nvmId = 0;
+    whNvmId keyCount;
+    /* try every index until we find a unique one, don't worry about capacity */
+    for (id = 1; id < WOLFHSM_KEYID_MASK; id++) {
+        /* check against cache keys */
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if ((id | WOLFHSM_KEYID_CRYPTO) == server->cache[i].meta->id)
+                break;
+        }
+        /* try again if match */
+        if (i < WOLFHSM_NUM_RAMKEYS)
+            continue;
+        /* if keyId exists */
+        ret = wh_Nvm_List(server->nvm, WOLFHSM_NVM_ACCESS_ANY,
+            WOLFHSM_NVM_FLAGS_ANY, id | WOLFHSM_KEYID_CRYPTO, &keyCount,
+            &nvmId);
+        /* break if we didn't find a match */
+        if (ret == WH_ERROR_NOTFOUND || nvmId != id)
+            break;
+    }
+    /* unlikely but cover the case where we've run out of ids */
+    if (id >= WOLFHSM_KEYID_MASK)
+        ret = WH_ERROR_NOSPACE;
+    /* ultimately, return found id */
+    if (ret == 0)
+        ret = (id | WOLFHSM_KEYID_CRYPTO);
+    return ret;
+}
+
+int hsmCacheKey(whServerContext* server, whNvmMetadata* meta, uint8_t* in)
+{
+    int i;
+    int foundIndex = -1;
+    /* make sure id is valid */
+    if (server == NULL || meta == NULL || in == NULL ||
+        meta->id == WOLFHSM_ID_ERASED ||
+        meta->len > WOLFHSM_NVM_MAX_OBJECT_SIZE) {
+        return WH_ERROR_BADARGS;
+    }
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* check for empty slot or rewrite slot */
+        if ((foundIndex == -1 &&
+            server->cache[i].meta->id == WOLFHSM_ID_ERASED) ||
+            server->cache[i].meta->id == meta->id) {
+            foundIndex = i;
+        }
+    }
+    /* if no empty slots, check for a commited key we can evict */
+    if (foundIndex == -1) {
+        for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+            if (server->cache[i].commited == 1) {
+                foundIndex = i;
+                break;
+            }
+        }
+    }
+    /* return error if we are out of cache slots */
+    if (foundIndex == -1)
+        return WH_ERROR_NOSPACE;
+    server->cache[foundIndex].commited = 0;
+    /* write key if slot found */
+    XMEMCPY((uint8_t*)server->cache[foundIndex].buffer, in, meta->len);
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(whNvmMetadata));
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* if this was RAM_KEY, set the global so we can export */
+    if (meta->id == WOLFHSM_SHE_TRANSLATE_KEY_ID(WOLFHSM_SHE_RAM_KEY_ID))
+        hsmSheRamKeyPlain = 1;
+#endif
+    return 0;
+}
+
+int hsmReadKey(whServerContext* server, whNvmMetadata* meta, uint8_t* out,
```

Comment:
This API conflates the NvmId and KeyId, which should be distinct.  Recommend to change this API to:
int hsmReadKey(whServerContext* server, whKeyId key_id, whNvmMetadata *out_meta, whNvmSize *inout_DataSize, uint8_t* data);

out_meta == NULL means don't copy the metadata.  inout_DataSize should be set to the length of data and will be set to the actual length if non-null.  data == NULL means don't copy the data.

I updated the key_id backlog item to capture the need for distinct key_id functions.

Reply:
I'm confused, are KeyIds not a subset of NvmId? It's just an NvmId but with the context of being a key and with either `WOLFHSM_KEYID_CRYPTO`, `WOLFHSM_KEYID_SHE` or `WOLFHSM_KEYID_SHE_RAM` bits set? Or is there something I'm missing